readability edit of section 4.0, complementing my ongoing draft review (#1014)
ietf-tapswg/api-drafts: draft-ietf-taps-impl.md

The process of establishing a network connection begins when an application expresses intent to communicate with a Remote Endpoint by calling Initiate. (At this point, any constraints or requirements the application may have on the connection are available from pre-establishment.) The process can be considered complete once there is at least one Protocol Stack that has completed any required setup to the point that it can transmit and receive the application's data.

> Connection establishment is divided into two top-level steps: Candidate Gathering, to identify the potential paths, protocols, and endpoints to use, and Candidate Racing (see Section 4.2.2 of {{I-D.ietf-taps-arch}}), in which the necessary protocol handshakes are conducted so that the transport system can select which set to use.

Reviewer: I think this may be more clear as "possible" or "available" than "potential"

> In its most basic form, the connection-establishment process might just involve identifying a single IP address (and port) to which the application wishes to connect, and starting a TCP handshake to establish a stream to that IP address, using the system's current default path (i.e., by using the current default network interface). However, each step may also differ depending on the requirements for the connection: if, instead of a known IP address, the endpoint is identified by a hostname, then there may be a choice between multiple resolved addresses; some protocols may not need any transport handshake to establish communication (such as UDP), while others may require several layers of handshakes, such as TLS over TCP; finally, there may also be multiple paths available (i.e., via additional network interfaces besides the system default);

Reviewer: Nit: I think "connection-establishment" should just be "connection establishment".

More broadly, I don't love how this makes the default seem like connecting to an IP address — "if, instead of a known IP address...". The default is going to be higher-level endpoints, and that's a core tenet of the API.

Reviewer 2: Agreed. Perhaps replacing the start of that sentence with something like "more commonly, when the endpoint..." will solve this.

Author: We could change

> if, instead of a known IP address, the endpoint is identified by a hostname, then there may be a choice between multiple resolved addresses;

to simply state that

> an endpoint is more commonly identified by a hostname which may resolve into more than one IP address;

Author: I think Tommy and Anna wanted to emphasize that having a name resolution step is the more common case.

Reviewer 3: I see, I'm a little unsure this specific change is really needed... and it doesn't seem to read well, but if it is needed then please consider  /which may/that can/ ... and /resolve into/resolve to/

|-> [Endpoint: 2001:DB8::1.80]     [Interface: LTE]   [Protocol: TCP]
~~~~~~~~~~

> Any one of these sub-entries on the aggregate connection attempt would satisfy the original application intent. The concern of this section is the algorithm defining which of these "Candidate" options to try, when, and in what order.

Reviewer: ```suggestion
Any one of these sub-entries on the aggregate connection attempt would satisfy the original application intent. The concern of this section is the algorithm defining which of these options (referred to for the rest of this section as "candidates") to try, when, and in what order.
```

> During Candidate Gathering, an implementation first excludes all protocols and paths with Prohibited features as well as those which do not satisfy a Required property.

Reviewer: ```suggestion
During Candidate Gathering, an implementation first excludes all protocols and paths that match Prohibited properties or do not satisfy the Required properties.
```

Then, the remaining candidates will be ranked according to Preferred properties, Avoided properties, and possibly other criteria.

## Structuring Candidates as a Tree

> The considereration of multiple candidates in a gathering and racing process can be conceptually structured as a tree; this terminological convention is used throughout this document. While a tree structure is not the only way in which racing can be implemented, it does ease the illustration of how racing works. 

Reviewer: ```suggestion
The considereration of multiple candidates in a gathering and racing process can be conceptually structured as a tree; this terminological convention is used throughout this document. While a tree structure is not the only way in which racing can be implemented, it is a useful model to use to understand and illustrate connection racing.
```

Author: Doesn't
> a useful model to use to

sound slightly odd and redundant?

How about simply saying
> it can serve as a useful model to illustrate connection racing

Reviewer 4: I don't understand what's slightly odd and redundant about the first version, but I guess both will work. This is for @tfpauly to answer though...
